Update support story for external calls and add sample

AlexJerabek opened the pull request.

opbld31 commented:
Docs Build status updates of commit _[cb452e4](https://github.com/OfficeDev/office-scripts-docs/commits/cb452e4759c0d9ac072f02ac89088af2024bd2f8)_: 

### :warning: Validation status: warnings


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :warning:Warning | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) | [Details](#user-content-74df4a59-7a726cb4)
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |

<a id="74df4a59-7a726cb4"></a>
### [docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md)
 - **Line 36, Column 5**: **[Warning-[file-not-found](https://review.docs.microsoft.com/help/contribute/validation-ref/file-not-found?branch=main)]** `````Invalid file link: '../resources/scenarios/noaa-data-fetch/md'.`````
 - **Line 48, Column 3**: **[Warning-[file-not-found](https://review.docs.microsoft.com/help/contribute/validation-ref/file-not-found?branch=main)]** `````Invalid file link: '../resources/scenarios/noaa-data-fetch/md'.`````

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C12%5C29%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202012290007015232-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=6D45zqIevo%2BskH0uECx5EuAljAGRcgH%2FcpVyRKRPgLE%3D&st=2020-12-29T00%3A02%3A30Z&se=2021-01-29T00%3A07%3A30Z&sp=r).

If you see build warnings/errors with permission issues, it might be due to single sign-on (SSO) enabled on Microsoft's GitHub organizations. Please follow instructions [here](https://teams.microsoft.com/l/message/19:7ecffca1166a4a3986fed528cf0870ee@thread.skype/1590030602688?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&parentMessageId=1590030602688&teamName=Docs%20Support&channelName=General&createdTime=1590030602688) to re-authorize your GitHub account to Docs Build.

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

**Note:** Your PR may contain errors or warnings unrelated to the files you changed. This happens when external dependencies like GitHub alias, Microsoft alias, cross repo links are updated. Please use these [instructions](https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master&tabs=pr-validation-error-solution#q-why-do-i-get-pr-validation-result-with-cryptic-error-or-warning-messages-not-related-to-the-changes-i-make-in-my-working-branch) to resolve them.

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

opbld30 commented:
Docs Build status updates of commit _[666ca0d](https://github.com/OfficeDev/office-scripts-docs/commits/666ca0d2dfbb55c8911bcdf469a269928ace4d96)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C12%5C29%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202012290007431674-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=Sky071sItcBlfwXvvFmn8UoIWLywla4CigJTPO2w3T4%3D&st=2020-12-29T00%3A03%3A13Z&se=2021-01-29T00%3A08%3A13Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

opbld30 commented:
Docs Build status updates of commit _[61d1d36](https://github.com/OfficeDev/office-scripts-docs/commits/61d1d36ffaa362d8446cc0a2f554e9950232884f)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C12%5C29%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202012290011379602-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=AHOvvV91xbXOFgCc3uXSvHSx0x%2FA%2Bx6wzEaOe13b6OM%3D&st=2020-12-29T00%3A07%3A16Z&se=2021-01-29T00%3A12%3A16Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

gergzk commented on docs/develop/external-calls.md:
```diff
@@ -1,28 +1,29 @@
 # External API call support in Office Scripts
 
-The Office Scripts platform doesn't support calls to [external APIs](https://developer.mozilla.org/docs/Web/API). However, these calls can be run under the right circumstances. External calls can be only be made through the Excel client, not through Power Automate [under normal circumstances](#external-calls-from-power-automate).
+Script authors shouldn't expect consistent behavior when using [external APIs](https://developer.mozilla.org/docs/Web/API) during the platform's preview phase. This is due how the JavaScript runtime manages interacting with the workbook. The script may end before the API call completes (or its `Promise` is fully resolved). As such, do not rely on external APIs for critical script scenarios.
```
re: "may end before its Promise is fully resolve" - the script ends when the promise resulting from main is resolved (do we want to be this specific? maybe), and is equally true for external and office js calls (though sync hides that in the officejs case). 

for external apis, the user needs to mark their main as async and await calls, it'll work as expected, eg:

async function main(workbook: ExcelScript.Workbook) {
    workbook.getActiveCell().setValue("Starting");
    const r = await fetch('https://jsonplaceholder.typicode.com/todos/1');
    const j = await r.json()
    workbook.getActiveWorksheet().getRange("A1").setValue(JSON.stringify(j));
    workbook.getActiveCell().setValue("Done");
}

However, the below would have non-deterministic behavior - the Done is likely to happen before the fetch completes, and main would exit and therefore the .then() wouldn't process.

function main(workbook: ExcelScript.Workbook) {
    workbook.getActiveCell().setValue("Starting");
fetch('https://jsonplaceholder.typicode.com/todos/1').then(r => r.json()).then(j =>  workbook.getActiveWorksheet().getRange("A1").setValue(JSON.stringify(j)));
    workbook.getActiveCell().setValue("Done");
}

gergzk commented on docs/develop/external-calls.md:
```diff
@@ -1,28 +1,29 @@
+Script authors shouldn't expect consistent behavior when using [external APIs](https://developer.mozilla.org/docs/Web/API) during the platform's preview phase. This is due how the JavaScript runtime manages interacting with the workbook. The script may end before the API call completes (or its `Promise` is fully resolved). As such, do not rely on external APIs for critical script scenarios.
 
-Script authors shouldn't expect consistent behavior when using external APIs during the platform's preview phase. This is due how the JavaScript runtime manages interacting with the workbook. The script may end before the API call completes (or its `Promise` is fully resolved). As such, do not rely on external APIs for critical script scenarios.
+Calls to external APIs can be only be made through the Excel client, not through Power Automate [under normal circumstances](#external-calls-from-power-automate).
```
Iirc you can make fetch calls in the PA AzFx runner, though there were some details there around admin lockdown. Please check with Yutao or Danny.

gergzk commented on docs/develop/external-calls.md:
```diff
@@ -1,28 +1,29 @@
+## Working with `fetch`
 
-The definition files for external APIs aren't included with Office Scripts. The use of such APIs generates compile-time errors for missing definitions. The APIs still run (though only when run through the Excel client), as shown in the following script:
+The [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) retrieves information from external services. It is an `async` API, so you will need to adjust the `main` signature of your script. Make the `main` function `async` and have it return a `Promise<void>`.
```
re; '... return a Promise<void>'

This is correct in the basic case, though the return type may evolve in PA scenarios. I'll leave it to you if that's intentional here.

gergzk commented on docs/develop/external-calls.md:
```diff
@@ -1,28 +1,29 @@
-   * "Cannot find name 'fetch'".
-   * It will still run and return the JSON from the testing service.
+  /* 
+   * Retrieve JSON data from a test server.
```
I shuold have read further before commenting above :D

opbld33 commented:
Docs Build status updates of commit _[7cb8c89](https://github.com/OfficeDev/office-scripts-docs/commits/7cb8c890246f8978d74d08f042cc06c2386101a0)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C12%5C29%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202012292249308307-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=mRt%2BKrK7Kn5BCsF2QgX1niOL7BIz%2BC6f9niU0nvKE4w%3D&st=2020-12-29T22%3A45%3A01Z&se=2021-01-29T22%3A50%3A01Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

opbld32 commented:
Docs Build status updates of commit _[bbbf68a](https://github.com/OfficeDev/office-scripts-docs/commits/bbbf68a30623dba3188d32e06639f0547a5782d0)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C12%5C29%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202012292250260930-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=v9UYnFZKydPklZsn%2FcxQp3PoiiDQ5EUqDQGF0LT4kmc%3D&st=2020-12-29T22%3A45%3A57Z&se=2021-01-29T22%3A50%3A57Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

AlexJerabek commented:
Thanks @gergzk. I think I address your feedback in the latest version. I'll work with the Power Automate crew to investigate using external APIs inside PA. For now, since it is known the basic scenario doesn't work, I'll that part of the documentation as is.

opbld33 commented:
Docs Build status updates of commit _[f6b76a2](https://github.com/OfficeDev/office-scripts-docs/commits/f6b76a2a07f1b23f1113b6e07574a027702600a4)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/toc.yml](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/toc.yml) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2021%5C1%5C4%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202101042334553087-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=xoXaebmI9KDkxXbk5FparZY1BCiM4zAkRbyIzR0vRFg%3D&st=2021-01-04T23%3A30%3A31Z&se=2021-02-04T23%3A35%3A31Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

alison-mk commented on docs/resources/scenarios/noaa-data-fetch.md:
```diff
@@ -0,0 +1,100 @@
+
+## Setup instructions
+
+1. Open the workbook with Excel for the web.
```
```suggestion
1. Open the workbook with Excel on the web.
```

alison-mk commented on docs/resources/scenarios/noaa-data-fetch.md:
```diff
@@ -0,0 +1,100 @@
+
+1. Open the workbook with Excel for the web.
+
+1. Under the **Automate** tab, open **All Scripts**.
```
Since **All Scripts** is a button and menu item, per the [style guide](https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=26472) the verb should be 'select'. 
```suggestion
1. Under the **Automate** tab, select **All Scripts**.
```

alison-mk commented on docs/resources/scenarios/noaa-data-fetch.md:
```diff
@@ -0,0 +1,100 @@
+
+1. Under the **Automate** tab, open **All Scripts**.
+
+1. In the **Code Editor** task pane, press **New Script** and paste the following script into the editor.
```
Also a button. Another reason not to use 'press' is that it is less inclusive -- it is mouse-centric and implies one is not using keyboard or voice navigation. 
```suggestion
1. In the **Code Editor** task pane, select **New Script** and paste the following script into the editor.
```

alison-mk commented on docs/resources/scenarios/noaa-data-fetch.md:
```diff
@@ -0,0 +1,100 @@
+      headerRange.getFormat().getFill().setColor("#4472C4");
+      headerRange.getFormat().getFont().setColor("white");
+    
+      // Insert all the data in rows from JSON
```
```suggestion
      // Insert all the data in rows from JSON.
```

alison-mk commented on docs/resources/scenarios/noaa-data-fetch.md:
```diff
@@ -0,0 +1,100 @@
+
+## Running the script
+
+On any worksheet, run the **NOAA Water Level Chart** script. The script fetches the water level data from December 25th 2020 to December 27th 2020. The `const` variables at the beginning of the script can be changed to use different dates or get different station information. The [CO-OPS API For Data Retrieval](https://api.tidesandcurrents.noaa.gov/api/prod/) describes how to get all this data.
```
Per [style guide](https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=26913). 
```suggestion
On any worksheet, run the **NOAA Water Level Chart** script. The script fetches the water level data from December 25, 2020 to December 27, 2020. The `const` variables at the beginning of the script can be changed to use different dates or get different station information. The [CO-OPS API For Data Retrieval](https://api.tidesandcurrents.noaa.gov/api/prod/) describes how to get all this data.
```

alison-mk reviewed (commented):
Love that sample! Added a handful of suggestions here. 

Plus one additional comment: This is not something you edited here, but I saw it while reviewing -- could you take a look at (new) line 43 in external-calls.md? Something is off about the first sentence in the !WARNING. I think there's a preposition missing.

opbld34 commented:
Docs Build status updates of commit _[f836cac](https://github.com/OfficeDev/office-scripts-docs/commits/f836cac1137701ca4171acbf10878a5e0af43aaf)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/toc.yml](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/toc.yml) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2021%5C1%5C5%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202101051809078753-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=LrZWqX%2B4X78Taj9G1OeTouoSeBFDaWY%2B1mVAuKPPtLM%3D&st=2021-01-05T18%3A04%3A39Z&se=2021-02-05T18%3A09%3A39Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

opbld32 commented:
Docs Build status updates of commit _[195bf73](https://github.com/OfficeDev/office-scripts-docs/commits/195bf73a73e7feb08b252728ce23c4c2e2be172e)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/toc.yml](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/toc.yml) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2021%5C1%5C5%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202101051822550290-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=4zyxvqVYqM2%2FLTcY%2B6YlOPR4Czakz7uTgQjiJg6b7u0%3D&st=2021-01-05T18%3A18%3A26Z&se=2021-02-05T18%3A23%3A26Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

AlexJerabek commented:
Thanks for the review @alison-mk. I've added your suggestions and adjusted the warning text that you called out. I also added a line to the script to insert a comment attributing the data's source, as per CELA's guidance.

opbld31 commented:
Docs Build status updates of commit _[5a93652](https://github.com/OfficeDev/office-scripts-docs/commits/5a936526abde501339cb87eaaffc1ae11854174b)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/external-calls.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/develop/external-calls.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |
[docs/images/scenario-noaa-water-level-after.PNG](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/images/scenario-noaa-water-level-after.PNG) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/resources/scenarios/noaa-data-fetch.md](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/resources/scenarios/noaa-data-fetch.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/resources/scenarios/noaa-data-fetch?branch=pr-en-us-204) |
[docs/toc.yml](https://github.com/OfficeDev/office-scripts-docs/blob/AlexJ-ExternalCalls/docs/toc.yml) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/office/dev/scripts/develop/external-calls?branch=pr-en-us-204) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2021%5C1%5C5%5C918d1c35-f3a4-631b-1edf-a1a2af22eee5%5CPullRequest%5C202101051938514905-204%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=UwzL64t%2B5WU90DnPZ89TFy%2FtVi3UaxyW3mPkjhBQJZQ%3D&st=2021-01-05T19%3A34%3A26Z&se=2021-02-05T19%3A39%3A26Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/918d1c35-f3a4-631b-1edf-a1a2af22eee5?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>
